### PR TITLE
[Role] `az role assignment create/update`: auto complete `assignee_principal_type`

### PR DESCRIPTION
**Pull request by evelyn-ys**
**Description**

Feature requested by #16006 
- When create role assignment, if `assignee` is provided without `assignee_principal_type`, try graph call to resolve object id and get object type
  - If graph call succeeds, resolve `object_id` and `principal_type`
  - If graph call fails, raise exception as before
- When create role assignment, if `assignee_object_id` is provided without `assignee_principal_type`, try to get principle type from graph.
  - If graph call succeeds, set `assignee_principal_type`
  - If graph call fails, leave `assignee_principal_type` as `None` as before

- When update role assignment, if `principal_type` is not provided, use the existing value of `assignment.principal_type`

**Testing Guide**

Graph call succeed:
```
> az role assignment create --role Reader --assignee-object-id "3707fb2f-ac10-4591-a04f-8b0d786ea37d"
{
  "canDelegate": null,
  "condition": null,
  "conditionVersion": null,
  "description": null,
  "id": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleAssignments/88670ac0-08a0-40af-8cb3-cfecc39224a6",
  "name": "88670ac0-08a0-40af-8cb3-cfecc39224a6",
  "principalId": "27c363a5-7016-4ae0-8540-818ec05673f1",
  "principalName": "",
  "principalType": "ServicePrincipal",
  "roleDefinitionId": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleDefinitions/acdd72a7-3385-48ef-bd42-f606fba81ae7",
  "roleDefinitionName": "Reader",
  "scope": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590",
  "type": "Microsoft.Authorization/roleAssignments"
}
```

Graph call fail:
```
> az role assignment create --role Reader --assignee-object-id "3707fb2f-ac10-4591-a04f-8b0d786ea37d"
WARNING: Failed to query --assignee-principal-type for 3707fb2f-ac10-4591-a04f-8b0d786ea37d by invoking Graph API.
RBAC server might reject creating role assignment by --assignee-object-id without --assignee-principal-type in the future. Better to specify --assignee-principal-type manually.
ERROR: The client '7701a65c-cdfe-4722-8cd7-a5dc211d87e6' with object id '7701a65c-cdfe-4722-8cd7-a5dc211d87e6' does not have authorization to perform action 'Microsoft.Authorization/roleAssignments/write' over scope '/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590' or the scope is invalid. If access was recently granted, please refresh your credentials.
```

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Role

File: src/azure-cli/azure/cli/command_modules/role/custom.py [hunk 1]
```diff
@@ -146,6 +146,16 @@ def create_role_assignment(cmd, role, assignee=None, assignee_object_id=None, re
...
 
+    # If assignee object id is provided without assignee principal type, try to get principle type from graph
+    if assignee_object_id and not assignee_principal_type:
+        graph_client = _graph_client_factory(cmd.cli_ctx)
+        try:
+            assignee_object_result = _get_object_stubs(graph_client, [assignee_object_id])
+            if assignee_object_result:
+                assignee_principal_type = assignee_object_result[0].object_type
+        except (CloudError, GraphErrorException):
+            pass
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/role/custom.py:157]: Please also make sure to test the scenario where the logged-in account doesn't have permission to query Graph API and gets `403` which is more frequent than receiving a `404`.

Comment (jiasli) [reply src/azure-cli/azure/cli/command_modules/role/custom.py:157]: Perhaps we should warn the user when graph query fails? as server may reject `"principalType": none` in the future.

Comment (evelyn-ys) [reply src/azure-cli/azure/cli/command_modules/role/custom.py]: `403` error example provided in PR description. Also added warning message 😊

File: src/azure-cli/azure/cli/command_modules/role/custom.py [hunk 2]
```diff
@@ -146,6 +146,19 @@ def create_role_assignment(cmd, role, assignee=None, assignee_object_id=None, re
     if assignee_principal_type and not assignee_object_id:
         raise CLIError('usage error: --assignee-object-id GUID [--assignee-principal-type]')
 
+    # If assignee object id is provided without assignee principal type, try to get principle type from graph
+    if assignee_object_id and not assignee_principal_type:
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/role/custom.py:150]: Per my understanding, the requirement of #16006 and email "Azure RBAC - Support passing attribute principalType during PUT operations HTTP calls" is to pass `principalType` whenever possible, not only when `--assignee-object-id` is provided.

In other words, this implementation passes `principalType` only when `--assignee-object-id` is provided. If `--assignee` is provided (user, group, service principal), `principalType` will not be passed.

Besides, the value assigned in `--assignee` will be converted to object ID anyway:

https://github.com/Azure/azure-cli/blob/5544a0f3bc32aa778d296f71c4d37e48f8ebe691/src/azure-cli/azure/cli/command_modules/role/custom.py#L1783

Comment (evelyn-ys) [reply src/azure-cli/azure/cli/command_modules/role/custom.py]: Resolved by
replacing `_resolve_object_id` in `_create_role_assignment`
with `_resolve_assignee_object` in `create_role_assignment`.

The latter will process differently between `assignee` and `assignee_object_id` and return both `object_id` and `object_type`

**Comment by jiasli** [comment]
Other references of `_create_role_assignment` should also be updated to provide `assignee_principal_type`, like

https://github.com/Azure/azure-cli/blob/5544a0f3bc32aa778d296f71c4d37e48f8ebe691/src/azure-cli/azure/cli/command_modules/role/custom.py#L1482

File: src/azure-cli/azure/cli/command_modules/role/custom.py [hunk 3]
```diff
@@ -1780,6 +1785,50 @@ def _encode_custom_key_description(key_description):
...
+                filter="servicePrincipalNames/any(c:c eq '{}')".format(assignee)))
+        if result:
+            return result[0].object_id, result[0].object_type
+
+    # verify object id
+    try:
+        if assignee_object_id:
+            result = _get_object_stubs(client, [assignee_object_id])
+        elif is_guid(assignee):
+            result = _get_object_stubs(client, [assignee])
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/role/custom.py:1807]: `fallback_to_object_id` by default is `False`. 

https://github.com/Azure/azure-cli/blob/5544a0f3bc32aa778d296f71c4d37e48f8ebe691/src/azure-cli/azure/cli/command_modules/role/custom.py#L1783

Using `assignee` as object ID causes a breaking change.

**Review by jiasli — APPROVED** [review]

**Review by fengzhou-msft — APPROVED** [review]

**Pull request merged**